salt/utils/process.py

```diff
@@ -195,71 +196,113 @@ def notify_systemd():
+
+    # all OSs supported by salt has psutil
+    if not HAS_PSUTIL:
+        log.critical(
+            "Assuming no other Process has this responsibly! pidfile: {}".format(
+                file_name
+            )
         )
-        sys.exit(salt.defaults.exitcodes.EX_NOUSER)
+        return True
 
-    if os.getuid() == uid:
-        # The current user already owns the pidfile. Return!
-        return
+    # add file directory if missing
+    file_directory_name = os.path.dirname(file_name)
+    if not os.path.isdir(file_directory_name) and file_directory_name:
+        os.makedirs(file_directory_name)
 
+    # get process info from file
+    file_process_info = None
     try:
-        os.chown(pidfile, uid, gid)
-    except OSError as err:
-        msg = "Failed to set the ownership of PID file {0} to user {1}.".format(
-            pidfile, user
-        )
-        log.debug("%s Traceback follows:", msg, exc_info=True)
-        sys.stderr.write("{0}\n".format(msg))
-        sys.exit(err.errno)
-    log.debug("Chowned pidfile: %s to user: %s", pidfile, user)
+        with salt.utils.files.fopen(file_name, "r") as file:
+            file_process_info = json.load(file)
+    except json.decoder.JSONDecodeError:
+        log.error("pidfile:{} is corrupted".format(file_name))
+    except FileNotFoundError:
+        log.info("pidfile: {} not found".format(file_name))
 
+    this_process_info = get_process_info()
 
-def check_pidfile(pidfile):
-    """
-    Determine if a pidfile has been written out
-    """
-    return os.path.isfile(pidfile)
+    # check if this process all ready has the responsibility
+    if file_process_info == this_process_info:
+        return True
+
+    # check if process is still alive
+    if file_process_info is not None and file_process_info == get_process_info(
+        file_process_info["pid"]
+    ):
+        return False
 
+    # process can take the mantle of responsibility
+    with salt.utils.files.fopen(file_name, "w") as file:
+        json.dump(this_process_info, file)
+    return True
 
-def get_pidfile(pidfile):
+
+def check_mantle_of_responsibility(file_name):
     """
-    Return the pid from a pidfile as an integer
+    Sees who has the mantle of responsibility
+    file_name: str
+    Return: None or int
     """
+
+    # all OSs supported by salt has psutil
+    if not HAS_PSUTIL:
+        log.critical(
+            "Assuming no other Process has this responsibly! pidfile: {}".format(
+                file_name
+            )
+        )
+        return
+
+    # get process info from file
     try:
-        with salt.utils.files.fopen(pidfile) as pdf:
-            pid = pdf.read().strip()
-        return int(pid)
-    except (OSError, IOError, TypeError, ValueError):
-        return -1
+        with salt.utils.files.fopen(file_name, "r") as file:
+            file_process_info = json.load(file)
+    except json.decoder.JSONDecodeError:
+        log.error("pidfile:{} is corrupted".format(file_name))
```

Comment:
no space after `:` - messages should be consistent.

Reply:
space